tool/tsh/app_cloud.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+func defaultValue[t any]() t {
+	var def t
+	return def
+}
+
+// pickCloudApp will attempt to find an active cloud app, automatically logging the user to the selected application if possible.
+func pickCloudApp[cloudApp any](cf *CLIConf, cloudFriendlyName string, matchRouteToApp func(tlsca.RouteToApp) bool, newCloudApp func(cf *CLIConf, profile *client.ProfileStatus, appRoute tlsca.RouteToApp) (cloudApp, error)) (cloudApp, error) {
+	app, needLogin, err := pickActiveCloudApp[cloudApp](cf, cloudFriendlyName, matchRouteToApp, newCloudApp)
+	if err != nil {
+		if !needLogin {
+			return defaultValue[cloudApp](), trace.Wrap(err)
+		}
+		log.WithError(err).Debugf("Failed to pick an active %v app, attempting to login into app %q", cloudFriendlyName, cf.AppName)
+		errLogin := onAppLogin(cf)
+		if errLogin != nil {
+			log.WithError(errLogin).Debugf("App login attempt failed")
+			// combine errors
+			var cApp cloudApp
+			return cApp, trace.NewAggregate(err, errLogin)
+		}
+		// another attempt
+		app, _, err = pickActiveCloudApp[cloudApp](cf, cloudFriendlyName, matchRouteToApp, newCloudApp)
+		return app, trace.Wrap(err)
+	}
+	return app, nil
+}
+
+func pickActiveCloudApp[cloudApp any](cf *CLIConf, cloudFriendlyName string, matchRouteToApp func(tlsca.RouteToApp) bool, newCloudApp func(cf *CLIConf, profile *client.ProfileStatus, appRoute tlsca.RouteToApp) (cloudApp, error)) (cApp cloudApp, needLogin bool, err error) {
+	profile, err := cf.ProfileStatus()
+	if err != nil {
+		return defaultValue[cloudApp](), false, trace.Wrap(err)
+	}
+	if len(profile.Apps) == 0 {
+		if cf.AppName == "" {
+			return defaultValue[cloudApp](), false, trace.NotFound("please login to %v app using 'tsh apps login' first", cloudFriendlyName)
+		}
+		return defaultValue[cloudApp](), true, trace.NotFound("please login to %v app using 'tsh apps login %v' first", cloudFriendlyName, cf.AppName)
```

Comment:
nit: `return defaultValue[cloudApp](), true, xxxx` cases probably don't need a user message anymore since a login will be done automatically?

Reply:
This can fail, in which case the message will be shown.